File: compiler/parser.py
from lexer import TokenType, Token
from ast_nodes import *

class ParseError(Exception):
    def __init__(self, message, token=None):
        super().__init__(message)
        self.message = message
        self.token = token

class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.pos = 0

    def current(self):
        return self.tokens[self.pos]

    def advance(self):
        self.pos += 1
        return self.tokens[self.pos-1]

    def expect(self, type, value=None):
        tok = self.current()
        if tok.type != type or (value is not None and tok.value != value):
            raise ParseError(f"Expected {type}{' '+value if value else ''}", tok)
        return self.advance()

    def parse(self):
        statements = []
        while self.current().type != TokenType.EOF:
            statements.append(self.statement())
        return Program(statements)

    def statement(self):
        tok = self.current()
        if tok.type == TokenType.KEYWORD and tok.value == 'dhori':
            return self.var_decl()
        if tok.type == TokenType.KEYWORD and tok.value == 'dekhao':
            self.advance()
            expr = self.expr()
            self.expect(TokenType.SEMI)
            return PrintStmt(expr)
        if tok.type == TokenType.KEYWORD and tok.value == 'jodi':
            return self.if_stmt()
        if tok.type == TokenType.KEYWORD and tok.value == 'ghuri':
            return self.for_stmt()
        # assignment statement: IDENT = expr;
        if tok.type == TokenType.IDENT:
            name = self.advance().value
            self.expect(TokenType.ASSIGN)
            expr = self.expr()
            self.expect(TokenType.SEMI)
            return AssignStmt(name, expr)
        raise ParseError("Unknown statement", tok)

    def var_decl(self):
        self.expect(TokenType.KEYWORD, 'dhori')
        name = self.expect(TokenType.IDENT).value
        self.expect(TokenType.ASSIGN)
        expr = self.expr()
        self.expect(TokenType.SEMI)
        return VarDecl(name, expr)

    def if_stmt(self):
        self.expect(TokenType.KEYWORD, 'jodi')
        cond = self.expr()
        self.expect(TokenType.LBRACE)
        then_stmts = []
        while self.current().type != TokenType.RBRACE:
            then_stmts.append(self.statement())
        self.expect(TokenType.RBRACE)
        else_branch = None
        if self.current().type == TokenType.KEYWORD and self.current().value in ('nahole', 'nahole-jodi'):
            self.advance()
            self.expect(TokenType.LBRACE)
            else_stmts = []
            while self.current().type != TokenType.RBRACE:
                else_stmts.append(self.statement())
            self.expect(TokenType.RBRACE)
            else_branch = else_stmts
        return IfStmt(cond, then_stmts, else_branch)

    def for_stmt(self):
        self.expect(TokenType.KEYWORD, 'ghuri')
        self.expect(TokenType.LPAREN)
        init = self.var_decl()
        cond = self.expr()
        self.expect(TokenType.SEMI)
        update_name = self.expect(TokenType.IDENT).value
        self.expect(TokenType.ASSIGN)
        update_expr = self.expr()
        self.expect(TokenType.RPAREN)
        self.expect(TokenType.LBRACE)
        body = []
        while self.current().type != TokenType.RBRACE:
            body.append(self.statement())
        self.expect(TokenType.RBRACE)
        return ForStmt(init, cond, (update_name, update_expr), body)

    def expr(self):
        left = self.term()
        while self.current().type in (TokenType.PLUS, TokenType.MINUS, TokenType.GT, TokenType.LT, TokenType.GE, TokenType.LE, TokenType.EQ):
            op = self.advance().type
            right = self.term()
            left = BinOp(left, op, right)
        return left

    def term(self):
        left = self.factor()
        while self.current().type in (TokenType.STAR, TokenType.SLASH):
            op = self.advance().type
            right = self.factor()
            left = BinOp(left, op, right)
        return left

    def factor(self):
        tok = self.current()
        if tok.type == TokenType.NUMBER:
            self.advance()
            return Num(tok.value)
        if tok.type == TokenType.IDENT:
            self.advance()
            return Var(tok.value)
        if tok.type == TokenType.LPAREN:
            self.advance()
            node = self.expr()
            self.expect(TokenType.RPAREN)
            return node
        raise ParseError("Expected expression", tok)

File: compiler/parser.ts
import { Token, TokenType, Lexer } from './lexer';
import {
  Program,
  Statement,
  VarDecl,
  AssignStmt,
  PrintStmt,
  IfStmt,
  ForStmt,
  Expression,
  BinOp,
  Num,
  Var,
} from './ast_nodes';

export class Parser {
  private tokens: Token[];
  private pos: number = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  private current(): Token {
    return this.pos < this.tokens.length ? this.tokens[this.pos] : this.tokens[this.tokens.length - 1];
  }

  private advance(): void {
    this.pos++;
  }

  private expect(type: TokenType): Token {
    const token = this.current();
    if (token.type !== type) {
      throw new Error(`Expected ${type} but got ${token.type}`);
    }
    this.advance();
    return token;
  }

  parse(): Program {
    const statements: Statement[] = [];
    while (this.current().type !== TokenType.EOF) {
      statements.push(this.statement());
    }
    return new Program(statements);
  }

  private statement(): Statement {
    if (this.current().type === TokenType.KEYWORD) {
      const keyword = this.current().value;
      if (keyword === 'dhori') {
        return this.varDecl();
      }
      if (keyword === 'dekhao') {
        return this.printStmt();
      }
      if (keyword === 'jodi') {
        return this.ifStmt();
      }
      if (keyword === 'ghuri') {
        return this.forStmt();
      }
    }
    if (this.current().type === TokenType.IDENT) {
      return this.assignStmt();
    }
    throw new Error(`Unknown statement: ${this.current()}`);
  }

  private varDecl(): VarDecl {
    this.expect(TokenType.KEYWORD); // dhori
    const name = this.expect(TokenType.IDENT).value;
    this.expect(TokenType.ASSIGN);
    const value = this.expr();
    this.expect(TokenType.SEMI);
    return new VarDecl(name, value);
  }

  private assignStmt(): AssignStmt {
    const name = this.expect(TokenType.IDENT).value;
    this.expect(TokenType.ASSIGN);
    const value = this.expr();
    this.expect(TokenType.SEMI);
    return new AssignStmt(name, value);
  }

  private printStmt(): PrintStmt {
    this.expect(TokenType.KEYWORD); // dekhao
    const value = this.expr();
    this.expect(TokenType.SEMI);
    return new PrintStmt(value);
  }

  private ifStmt(): IfStmt {
    this.expect(TokenType.KEYWORD); // jodi
    const condition = this.expr();
    this.expect(TokenType.LBRACE);
    const thenBlock: Statement[] = [];
    while (this.current().type !== TokenType.RBRACE) {
      thenBlock.push(this.statement());
    }
    this.expect(TokenType.RBRACE);

    let elseBlock: Statement[] | null = null;
    if (this.current().type === TokenType.KEYWORD && this.current().value === 'nahole') {
      this.expect(TokenType.KEYWORD); // nahole
      this.expect(TokenType.LBRACE);
      elseBlock = [];
      while (this.current().type !== TokenType.RBRACE) {
        elseBlock.push(this.statement());
      }
      this.expect(TokenType.RBRACE);
    }

    return new IfStmt(condition, thenBlock, elseBlock);
  }

  private forStmt(): ForStmt {
    this.expect(TokenType.KEYWORD); // ghuri
    this.expect(TokenType.LPAREN);
    const init = this.varDecl(); // This consumes the SEMI
    const condition = this.expr();
    this.expect(TokenType.SEMI);
    // For update: parse assignment without expecting SEMI at the end (we'll consume RPAREN next)
    const updateName = this.expect(TokenType.IDENT).value;
    this.expect(TokenType.ASSIGN);
    const updateValue = this.expr();
    const update = new AssignStmt(updateName, updateValue);
    this.expect(TokenType.RPAREN);
    this.expect(TokenType.LBRACE);
    const body: Statement[] = [];
    while (this.current().type !== TokenType.RBRACE) {
      body.push(this.statement());
    }
    this.expect(TokenType.RBRACE);
    return new ForStmt(init, condition, update, body);
  }

  private expr(): Expression {
    return this.binOp();
  }

  private binOp(minPrec: number = 0): Expression {
    let left = this.unary();

    while (this.isBinOp()) {
      const op = this.current().value;
      const prec = this.precedence(op);
      if (prec < minPrec) break;

      this.advance();
      const right = this.binOp(prec + 1);
      left = new BinOp(left, op, right);
    }

    return left;
  }

  private isBinOp(): boolean {
    const type = this.current().type;
    return (
      type === TokenType.PLUS ||
      type === TokenType.MINUS ||
      type === TokenType.STAR ||
      type === TokenType.SLASH ||
      type === TokenType.GT ||
      type === TokenType.LT ||
      type === TokenType.GE ||
      type === TokenType.LE ||
      type === TokenType.EQ
    );
  }

  private precedence(op: string): number {
    switch (op) {
      case '+':
      case '-':
        return 1;
      case '*':
      case '/':
        return 2;
      case '>':
      case '<':
      case '>=':
      case '<=':
      case '==':
        return 0;
      default:
        return -1;
    }
  }

  private unary(): Expression {
    if (this.current().type === TokenType.MINUS) {
      this.advance();
      return new BinOp(new Num(0), '-', this.unary());
    }
    return this.primary();
  }

  private primary(): Expression {
    if (this.current().type === TokenType.NUMBER) {
      const value = this.current().value;
      this.advance();
      return new Num(value);
    }

    if (this.current().type === TokenType.IDENT) {
      const name = this.current().value;
      this.advance();
      return new Var(name);
    }

    if (this.current().type === TokenType.LPAREN) {
      this.advance();
      const expr = this.expr();
      this.expect(TokenType.RPAREN);
      return expr;
    }

    throw new Error(`Unexpected token: ${this.current()}`);
  }
}
